[@backstage/cli]: Unable to run a project with only backend plugins due to missing "react" dependency

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior

`yarn start` in a project that's only using backend plugins (a `proxy` or `auth` service for example) should work.

<!--- Tell us what should happen -->

## Current Behavior

<!--- Tell us what happens instead of the expected behavior -->

`yarn start` is failing due to a missing dependency for `react`:

```
Error: Cannot find module 'react'
...
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'node_modules/react-hot-loader/dist/react-hot-loader.production.min.js',
    'node_modules/react-hot-loader/index.js',
    'packages/backend/dist/main.js'
  ]
```

## Possible Solution

<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas as to the implementation of the addition or change -->

Seems that this was introduced in `v0.10.1` of `@backstage/cli` as I was able to successfully run the service using `v0.10.0`.

Adding `react` to `devDependencies` works as a work-around for now.

## Steps to Reproduce

<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code or configuration to reproduce, if relevant -->

1. Use a project with only backend plugins (no frontend app)
2. Use `"@backstage/cli": "^0.10.1`
3. `yarn start`

## Context

<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context (e.g. links to configuration settings, -->
<!--- stack trace or log data) helps us come up with a solution that is most useful in the real world -->

## Your Environment

<!--- Include as many relevant details about the environment you experienced the bug in -->

<!-- ProTip: You can use `yarn backstage-cli info` command in your Backstage App for this section. -->

- NodeJS Version (v14):
- Operating System and Version (e.g. Ubuntu 14.04):
- Browser Information:

```
OS:   Darwin 19.6.0 - darwin/x64
node: v16.10.0
yarn: 1.22.10
```

--- a/.changeset/quick-avocados-sell.md
+++ b/.changeset/quick-avocados-sell.md
@@ -1,0 +1,5 @@
+---
+'@backstage/cli': patch
+---
+
+Fixed a bug were the `react-hot-loader` transform was being applied to backend development builds.

--- a/packages/cli/src/lib/bundler/config.ts
+++ b/packages/cli/src/lib/bundler/config.ts
@@ -226,7 +226,7 @@ export async function createBackendConfig(
   // See frontend config
   const externalPkgs = packages.filter(p => !isChildPath(paths.root, p.dir));
 
-  const { loaders } = transforms(options);
+  const { loaders } = transforms({ ...options, isBackend: true });
 
   const runScriptNodeArgs = new Array<string>();
   if (options.inspectEnabled) {

--- a/packages/cli/src/lib/bundler/transforms.ts
+++ b/packages/cli/src/lib/bundler/transforms.ts
@@ -25,12 +25,13 @@ type Transforms = {
 
 type TransformOptions = {
   isDev: boolean;
+  isBackend?: boolean;
 };
 
 export const transforms = (options: TransformOptions): Transforms => {
-  const { isDev } = options;
+  const { isDev, isBackend } = options;
 
-  const extraTransforms = isDev ? ['react-hot-loader'] : [];
+  const extraTransforms = isDev && !isBackend ? ['react-hot-loader'] : [];
 
   // This ensures that styles inserted from the style-loader and any
   // async style chunks are always given lower priority than JSS styles.

